**Проект по SQL для сервиса по прочтению книг.**

Цель — проанализировать базу данных крупного сервиса для чтения книг по подписке и сформулировать ценностное предложение для нового продукта. В базе данных — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

Описание данных

Таблица **books** Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица **authors** Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

Таблица **publishers** Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица **ratings** Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица **reviews** Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

Задания:

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Структура данных:**


<img src=https://storage.yandexcloud.net/cloud-evseenko/Untitled_(33).png>
<br>


# Предобработка

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'],
                                                        db_config['db']) 

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [3]:
#изучим таблицы

info_func = 'SELECT * FROM {}'

data = ['books', 'authors', 'publishers', 'ratings', 'reviews']

#в цикле выведем первые 5 строк, инфо, пропуски и дубликаты
for i in data:
    print('\033[1m' + i + '\033[0m')
    display(pd.read_sql(info_func.format(i), engine).head())
    print()
    print('\033[1m' + 'метод info' + '\033[0m')
    display((pd.read_sql(info_func.format(i), engine).info()))
    print()
    print('\033[1m' + 'метод isna().sum()' + '\033[0m')
    display((pd.read_sql(info_func.format(i), engine).isna().sum()))
    print()
    print('\033[1m' + 'метод .duplicated().sum()' + '\033[0m')
    display((pd.read_sql(info_func.format(i), engine).duplicated().sum()))
    print()

books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


метод info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None


метод isna().sum()


book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64


метод .duplicated().sum()


0


authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


метод info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None


метод isna().sum()


author_id    0
author       0
dtype: int64


метод .duplicated().sum()


0


publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


метод info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None


метод isna().sum()


publisher_id    0
publisher       0
dtype: int64


метод .duplicated().sum()


0


ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


метод info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None


метод isna().sum()


rating_id    0
book_id      0
username     0
rating       0
dtype: int64


метод .duplicated().sum()


0


reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


метод info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None


метод isna().sum()


review_id    0
book_id      0
username     0
text         0
dtype: int64


метод .duplicated().sum()


0

Дубликатов, пропусков нет, в целом данные в порядке, можно продолжать анализ.

## 1 задание: Посчитайте, сколько книг вышло после 1 января 2000 года

In [4]:
#функция для выполнения sql-запроса
def result(query):
    return pd.io.sql.read_sql(query, con = engine)

In [5]:
query1 = '''SELECT COUNT(title)
            FROM books
            WHERE publication_date >= '2000-01-01'
         '''

print('После 1 января 2000 года вышло {} книг'.format(result(query1).values[0]))

После 1 января 2000 года вышло [821] книг


## 2 задание: Для каждой книги посчитайте количество обзоров и среднюю оценку

In [7]:
query2 =  '''SELECT books.*, COUNT(distinct review_id) as reviews_num, AVG(ratings.rating) as avg_rating
             FROM  books
             LEFT JOIN reviews ON books.book_id = reviews.book_id
             LEFT JOIN ratings ON books.book_id = ratings.book_id
             GROUP BY books.book_id
             ORDER BY reviews_num  DESC 
             
        '''


books_rev_rtg = result(query2)

print('Таблица с количством обзоров и средней оценкой книг (сортировка по количеству обзоров)')
display(books_rev_rtg.head(10))

Таблица с количством обзоров и средней оценкой книг (сортировка по количеству обзоров)


book_id  author_id                                              title  \
0      948        554                            Twilight (Twilight  #1)   
1      963        527                                Water for Elephants   
2      734        278                                   The Glass Castle   
3      302        236  Harry Potter and the Prisoner of Azkaban (Harr...   
4      695        395  The Curious Incident of the Dog in the Night-Time   
5      696        106             The Da Vinci Code (Robert Langdon  #2)   
6      627        469                                      The Alchemist   
7      750        240                The Hobbit  or There and Back Again   
8      656        402                                     The Book Thief   
9      779        498  The Lightning Thief (Percy Jackson and the Oly...   

   num_pages publication_date  publisher_id  reviews_num  avg_rating  
0        501       2006-09-06           176            7    3.662500  
1        335       2007-05-01            12            6    3.977273  
2        288       2006-01-17           258            6    4.206897  
3        435       2004-05-01           255            6    4.414634  
4        226       2004-05-18           309            6    4.081081  
5        489       2006-03-28            15            6    3.830508  
6        197       1993-05-01           133            6    3.789474  
7        366       2002-08-15           154            6    4.125000  
8        552       2006-03-14             9            6    4.264151  
9        375       2006-03-01            91            6    4.080645

## 3 задание: Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query3 = '''SELECT publishers.*, COUNT(book_id) as book_cnt
           FROM 
                  publishers 
                  INNER JOIN books ON publishers.publisher_id = books.publisher_id
           WHERE num_pages > 50      
           GROUP BY publishers.publisher_id
           ORDER BY COUNT(book_id) DESC
           LIMIT 1
        '''

print('Издательство, которое выпустило наибольшее число книг толще 50 страниц:')
display(result(query3))

Издательство, которое выпустило наибольшее число книг толще 50 страниц:


publisher_id      publisher  book_cnt
0           212  Penguin Books        42

## 4 задание: Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [10]:
query4 = '''SELECT author, AVG(subq.avg_rating) AS avg_rtg
            FROM
                (SELECT author, books.book_id, AVG(rating) AS avg_rating
                FROM authors
                LEFT JOIN books ON books.author_id = authors.author_id
                LEFT JOIN ratings ON ratings.book_id = books.book_id
                GROUP BY author, books.book_id
                HAVING COUNT(rating) > 50) AS subq
            GROUP BY author
            ORDER BY avg_rtg DESC
            LIMIT 1
           
           
        '''

print('Автор с самой высокой средней оценкой книг (с 50 оценками и более):')
display(result(query4))

Автор с самой высокой средней оценкой книг (с 50 оценками и более):


author   avg_rtg
0  J.K. Rowling/Mary GrandPré  4.283844

## 5 задание: Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [11]:
query5 = '''SELECT AVG(reviews_num)
           FROM 
               (SELECT ratings.username, COUNT(DISTINCT review_id) as reviews_num
                FROM 
                reviews
                RIGHT OUTER JOIN ratings ON reviews.username = ratings.username         
                GROUP BY ratings.username
                HAVING COUNT(DISTINCT ratings.book_id) > 50) as A
        '''

print('Среднее количество обзоров от пользователей, которые поставили больше 50 оценок: ',result(query5).values[0].round(1))

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок:  [24.3]


## Общие выводы

- После 1 января 2000 года вышло 821 книг
- топ 5 книг по количеству обзоров, а значит по общественному интересу:
    - Twilight (Twilight #1)
    - The Hobbit or There and Back Again
    - The Catcher in the Rye
    - Harry Potter and the Prisoner of Azkaban
    - Harry Potter and the Chamber of Secrets
- Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books (42 книги)
- Автор с самой высокой средней оценкой книг (с 50 оценками и более) - J.K. Rowling/Mary GrandPré (ср.рейтинг 4.3)
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок:  24.3

Чтобы на старте завоевать как можно больше аудитории, необходимо бить в самые популярные тренды, чтобы как можно быстрее выйти на достаточное количество подписчиков. Для этого рекомендуется:
- первоочередно продвигать популярные серии романов, отслеживать их можно по количеству обзоров пользовтелей - чем их больше, тем сильнее внимание аудитории к книге и серии в целом. 
- обратить внимание на издательство Penguin Books - они выпустили 42 книги, больше, чем все остальные издатели. Вероятно, у них есть какое-то конкурентное преимущество (может быть опыт, пиар и т.д.), которым можно воспользоваться.
- акцентировать внимание на авторов с высокими оценками книг, например J.K. Rowling/Mary GrandPré (ср.рейтинг 4.3), т.к. такие книги может выбирать тот сегмент аудитории, который делает свой выбор исходя из рейтингов (по аналогии с кино и ресторанами)
- мониторить "инфлюенсеров" среди книг - тех пользователей, которые делают много обзоров и ставят много оценок. Наверняка эти люди разбираются в том, что делают и имеют какую-то свою аудиторию. У таких пользователей есть смысл купить рекламу своего сервиса в их блогах или соцсетях, чтобы максимально точно охватить нашу целевую аудиторию.